src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs

```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
             });
         }
 
-        private void PartitionsAssigned(IEnumerable<TopicPartition> partitions)
+        private class FlushMessagesOfRevokedPartitionsHandler : IPartitionEventHandler
         {
-            TopicPartitions = TopicPartitions.Union(partitions);
-            Log.Debug("[{0}] Partitions were assigned: {1}", ConsumerActor.Path.Name, string.Join(", ", partitions));
-            RequestMessages();
+            private IImmutableSet<TopicPartitionOffset> _lastRevoked = ImmutableHashSet<TopicPartitionOffset>.Empty;
+            private readonly SingleSourceStageLogic<K, V, TMessage> _stageLogic;
+
+            public FlushMessagesOfRevokedPartitionsHandler(SingleSourceStageLogic<K, V, TMessage> stageLogic)
+            {
+                _stageLogic = stageLogic;
+            }
+
+            public void OnRevoke(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions,
+                IRestrictedConsumer consumer)
+            {
+                _lastRevoked = revokedTopicPartitions;
+            }
+
+            public void OnLost(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                _stageLogic.FilterRevokedPartitionAsyncCallback(revokedTopicPartitions);
+            }
+
+            public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                // remove all of our previous revoked partitions that are not in the new assignment
+                _stageLogic.FilterRevokedPartitionAsyncCallback(_lastRevoked
+                    .Where(c => !assignedTopicPartitions.Contains(c.TopicPartition))
+                        .ToImmutableHashSet());
+            }
+
+            public void OnStop(IImmutableSet<TopicPartition> topicPartitions, IRestrictedConsumer consumer){}
         }
-        
-        private void PartitionsRevoked(IEnumerable<TopicPartitionOffset> partitions)
+
+        protected override IPartitionEventHandler AddToPartitionAssignmentHandler(IPartitionEventHandler handler) => 
+           new PartitionEventHandlers.Chain(new FlushMessagesOfRevokedPartitionsHandler(this), handler);
+
+        private void PartitionsAssigned(IImmutableSet<TopicPartition> partitions)
         {
-            TopicPartitions = TopicPartitions.Except(partitions.Select(tpo => tpo.TopicPartition));
-            Log.Debug("[{0}] Partitions were revoked: {1}", ConsumerActor.Path.Name, string.Join(", ", partitions));
+            TopicPartitions = TopicPartitions.Union(partitions);
+            Log.Debug("[{0}] Partitions were assigned: {1}. All partitions: {2}", ConsumerActor.Path.Name, string.Join(", ", partitions), string.Join(", ", TopicPartitions));
+            RequestMessages();
         }
         
-        private void PartitionsLost(IEnumerable<TopicPartitionOffset> partitions)
```

Comment:
Remove the separate `PartitionsLost` method and just consolidate it with `PartitionsRevoked` - it's not a meaningful difference to us.